lib/Sema/SemaBounds.cpp

```diff
@@ -628,93 +698,210 @@ namespace {
+            return CreateBoundsEmpty();
+
+          // `e1 = e2` has the bounds of `e2`. `e2` is an RValue.
           if (Op == BinaryOperatorKind::BO_Assign)
             return RValueBounds(RHS);
 
+          // `e1, e2` has the bounds of `e2`. Both `e1` and `e2`
+          // are RValues.
+          if (Op == BinaryOperatorKind::BO_Comma)
+            return RValueBounds(RHS);
+
+          // Compound Assignments function like assignments mostly,
+          // except the LHS is an L-Value, so we'll use its lvalue target bounds
           bool IsCompoundAssignment = false;
           if (BinaryOperator::isCompoundAssignmentOp(Op)) {
             Op = BinaryOperator::getOpForCompoundAssignment(Op);
             IsCompoundAssignment = true;
           }
 
           // Pointer arithmetic.
+          //
+          // `p + i` has the bounds of `p`. `p` is an RValue.
+          // `p += i` has the lvalue target bounds of `p`. `p` is an LValue.
+          // same applies for `-` and `-=` respectively
           if (LHS->getType()->isPointerType() &&
               RHS->getType()->isIntegerType() &&
               BinaryOperator::isAdditiveOp(Op)) {
             return IsCompoundAssignment ?
               LValueTargetBounds(LHS) : RValueBounds(LHS);
           }
+          // `i + p` has the bounds of `p`. `p` is an RValue.
+          // `i += p` has the bounds of `p`. `p` is an RValue.
           if (LHS->getType()->isIntegerType() &&
               RHS->getType()->isPointerType() &&
-              BinaryOperator::isAdditiveOp(Op)) {
-            assert(!IsCompoundAssignment);
+              Op == BinaryOperatorKind::BO_Add) {
             return RValueBounds(RHS);
           }
+          // `e - p` has empty bounds, regardless of the bounds of p.
+          // `e -= p` has empty bounds, regardless of the bounds of p.
+          if (RHS->getType()->isPointerType() &&
+              Op == BinaryOperatorKind::BO_Sub) {
+            return CreateBoundsEmpty();
+          }
 
           // Arithmetic on integers with bounds.
+          //
+          // `e1 @ e2` has the bounds of whichever of `e1` or `e2` has bounds.
+          // if both `e1` and `e2` have bounds, then they must be equal.
+          // Both `e1` and `e2` are RValues
+          //
+          // `e1 @= e2` has the bounds of whichever of `e1` or `e2` has bounds.
+          // if both `e1` and `e2` have bounds, then they must be equal.
+          // `e1` is an LValue, its bounds are the lvalue target bounds.
+          // `e2` is an RValue
+          //
+          // @ can stand for: +, -, *, /, %, &, |, ^, >>, <<
           if (LHS->getType()->isIntegerType() &&
               RHS->getType()->isIntegerType() &&
               (BinaryOperator::isAdditiveOp(Op) ||
                BinaryOperator::isMultiplicativeOp(Op) ||
-               BinaryOperator::isBitwiseOp(Op))) {
+               BinaryOperator::isBitwiseOp(Op) ||
+               BinaryOperator::isShiftOp(Op))) {
```

Comment:
Added shift operators here, because they're kinda bitwise, kinda multiplication, and it seems right to treat them like the other mathematical operators, rather than like a logical operator.